### PR TITLE
Swb CTC SCF: Audio perturbations

### PR DESCRIPTION
**Pull request by Max-Ryujin**
Searching over all combinations of speed, tempo, preemphasis, and codec including the baseline.
This replaces the previous `run_mel_audio_perturbation()` function, which only implemented a few cases.

### CODE REVIEW

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 1]
```diff
@@ -38,7 +38,8 @@ def get_nn_args(nn_base_args, num_epochs, evaluation_epochs=None, prefix="", tra
 
     for name, args in nn_base_args.items():
         returnn_config, returnn_recog_config, report_args = get_nn_args_single(
-            num_epochs=num_epochs, evaluation_epochs=evaluation_epochs, **copy.deepcopy(args))
```
Comment (Max-Ryujin) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:41]: Black formatting

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 2]
```diff
@@ -67,8 +68,13 @@ def get_nn_args(nn_base_args, num_epochs, evaluation_epochs=None, prefix="", tra
 
 
 def get_nn_args_single(
-    num_outputs: int = 88, num_epochs: int = 500, evaluation_epochs: Optional[List[int]] = None,
```
Comment (Max-Ryujin) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:70]: Black formatting

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 3]
```diff
@@ -67,8 +68,13 @@ def get_nn_args(nn_base_args, num_epochs, evaluation_epochs=None, prefix="", tra
...
-    lr_args=None, feature_args=None, returnn_args=None, report_args=None,
```
Comment (Max-Ryujin) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:71]: Black formatting

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 4]
```diff
@@ -171,11 +177,11 @@ def get_returnn_config(
         "phon_future_length": 0,
         "allophone_file": tk.Path(
             "/u/vieting/setups/swb/20230406_feat/dependencies/allophones_blank",
-            hash_overwrite="SWB_ALLOPHONE_FILE_WEI_BLANK"
```
Comment (Max-Ryujin) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:174]: Black formatting

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 5]
```diff
@@ -229,7 +235,7 @@ def get_returnn_config(
             "learning_rate_control_relative_error_relative_lr": True,
             "min_learning_rate": 1e-5,
             "newbob_learning_rate_decay": 0.9,
-            "newbob_multi_num_epochs": datasets["train"]["partition_epoch"],
```
Comment (Max-Ryujin) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:232]: This change is needed, because the dataset changed to a Multiproc Dataset

File: users/vieting/experiments/switchboard/ctc/feat/default_tools.py [hunk 6]
```diff
@@ -31,6 +31,6 @@
 
 RETURNN_ROOT = CloneGitRepositoryJob(
     "https://github.com/rwth-i6/returnn",
-    commit="6d2945a85cc95df5349a59541d84f172dd55cc20",
```
Comment (Max-Ryujin) [inline users/vieting/experiments/switchboard/ctc/feat/default_tools.py:34]: Updated to a newer Returnn commit. Was needed to solve some older Problems. 
Maybe change this part completely to not use fixed commits, but always the newest commit?

Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/default_tools.py:34]: I think it's better to have the hash in there. Also, you might change it for only a subset of the experiments if it is only needed there to make sure everything is reproducible.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 7]
```diff
@@ -48,7 +48,12 @@ def get_datasets(**kwargs):
     )
 
     returnn_datasets = {
-        "train": nn_train_data_inputs["switchboard.train"].get_data_dict()["datasets"]["ogg"],
```
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:51]: around 20%. epochs take around 50 minutes instead of an hour

[the remark this answers is not in the document]
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:51]: What is the compute time before and after?

[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py]: It is identical. Reverted back for simplicity

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 8]
```diff
@@ -744,7 +744,7 @@ def run_scf_baseline():
             ),
         },
         num_epochs=450,
-        prefix="conformer_bs2x5k_",
```
Comment (Max-Ryujin) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:747]: All batch sizes are moved to 5000 to make all future audio perturbation experiments small enough for 11GB cards

File: users/vieting/experiments/switchboard/ctc/feat/default_tools.py (hunk 6 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/default_tools.py]: reverted to old hash

**Comment by vieting** [comment]
Is the PR ready for review? Please resolve conflicts.

**Comment by Max-Ryujin** [comment]
I want to check if the audio perturbation and baseline train and converge, before closing that PR.

**Comment by vieting** [comment]
> I want to check if the audio perturbation and baseline train and converge, before closing that PR.

Alright. I converted the PR to a draft. Please mark as ready for review and request a review when you think it's ready.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 9]
```diff
@@ -719,41 +724,354 @@ def run_scf_baseline():
...
+                                    "ReturnnTrainingJob.y9otnVMrBAWw/output/models/backup.epoch.210"
+                                ),
+                                "init_for_train": True,
+                            },
+                        },
+                    },
+                },
+                feature_args=feature_args,
+                lr_args={**lr_args, "peak_lr": 3.35e-4, "increase_epochs": 0, "decrease_epochs": 150},
+                report_args={"batch_size": "2x5k"},
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:1074]: You copied this part from me apparently. I would prefer not to have it in this PR.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 10]
```diff
@@ -1,4 +1,5 @@
 from typing import List, Dict, Optional, Any
+import inspect
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:2]: Also here, these changes are not really what the title suggests this PR is about. Please split this up so that we can also have a somewhat helpful git history.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 11]
```diff
@@ -719,41 +724,354 @@ def run_scf_baseline():
...
+        key_suffix.append("speed" + f"{speed['prob']}_{speed['minimum']}_{speed['maximum']}")
+        key = f"scf_bs7k_conf-wei-oldspecaug-audio_perturbation_{'_'.join(key_suffix)}"
+        audio_perturb_args = {}
+        audio_perturb_args["speed"] = speed
+        report_args = {
+            "architecture": "conf-wei",
+            "lr": "wei_peak_4e-4_e450_cycle360",
+            "specaug": "wei_adapt_80dim",
+            "wave_norm": "True",
+        }
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:864]: I think if this is the baseline, you could also omit them here if you don't want to merge this report with the other reports that did use these args.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 12]
```diff
@@ -719,41 +724,354 @@ def run_scf_baseline():
...
+
+    nn_base_args = {}
+
+    for speed in speeds:
+        key_suffix = []
+        # build key suffix
+        key_suffix.append("speed" + f"{speed['prob']}_{speed['minimum']}_{speed['maximum']}")
+        key = f"scf_bs7k_conf-wei-oldspecaug-audio_perturbation_{'_'.join(key_suffix)}"
+        audio_perturb_args = {}
+        audio_perturb_args["speed"] = speed
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:858]:
```suggestion
        audio_perturb_args = {"speed": speed}
```

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 13]
```diff
@@ -915,11 +1236,13 @@ def py():
...
-        report_mel,
-        report_scf,
-    ])
+    report = Report.merge_reports(
+        [
+            report_base,
+            report_mel,
+            report_scf,
+        ]
+    )
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:1245]: Unrelated to actual PR.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 14]
```diff
@@ -719,41 +724,354 @@ def run_scf_baseline():
...
-                    "specaug": "wei_adapt_80dim",
-                    "wave_norm": "True",
+                **returnn_args,
+            },
+            feature_args=feature_args,
+            lr_args=lr_args,
+            report_args=report_args,
+        )
+
+    for tempo in tempos:
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:881]: Please check the comments I added for speed perturbation for all perturbations below. Also, there is a lot of code duplication below. I think you could solve this much more elegantly for the different perturbations.

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:232]: Yes. MultiProc Dataset is needed for audio perturbation and has no negative effects on other trainings.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:52]: With the one change in baseline args, it works for other trainings as well. But I can make it selectable as well.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 15]
```diff
@@ -719,41 +724,354 @@ def run_scf_baseline():
         dev_corpora,
     ) = get_datasets()
     returnn_args = {
-        "batch_size": 5000,
+        "batch_size": 7000,
```
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:727]: I changed the baseline from 2x5k to 7k to be in line with the audio perturbation experiments with 7k. 5k does not work, because the trainings do not converge.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py (hunk 12 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:858]: is this the recommended formating for all arguments, or are there cases where the [""] syntax is better?

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:232]: If you enable it for all trainings, it changes all hashes which is never a good idea. Please only use it when necessary.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:727]: But why not 2x5k? I'm not comfortable with changing the baseline unless there is evidence to do it.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py (hunk 12 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:858]: The `[""]` syntax is typically used when you modify an entry of an existing dict. When you initialize it, it is more concise to write it directly in there and I would always prefer that.

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 16]
```diff
@@ -142,6 +142,7 @@ def get_returnn_config(
     extra_args: Optional[Dict[str, Any]] = None,
     staged_opts: Optional[Dict[int, Any]] = None,
     audio_perturbation: bool = False,
+    useMultiProcDataset: bool = False,
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:145]: Please stay consistent with the style (no camel case).
```suggestion
    use_multi_proc_dataset: bool = False,
```

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 17]
```diff
@@ -215,6 +216,11 @@ def get_returnn_config(
     if isinstance(batch_size, int):
         # If batch size is int, adapt to waveform. If it is dict, assume it is already correct.
         batch_size = {"classes": batch_size, "data": batch_size * sample_rate // 100}
+    if useMultiProcDataset:
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:219]: See above

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:232]: This change here seems to be wrong, doesn't it?

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 18]
```diff
@@ -27,7 +27,7 @@
 from .default_tools import RASR_BINARY_PATH, RETURNN_ROOT, RETURNN_EXE, SCTK_BINARY_PATH
 
 
-def get_datasets(**kwargs):
+def get_datasets(**kwargs, useMultiProcDataset=False):
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:30]: See above.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 19]
```diff
@@ -47,14 +47,27 @@ def get_datasets(**kwargs):
...
+            },
+        }
+    else:
+        returnn_datasets = {
+            "train": nn_train_data_inputs["switchboard.train"].get_data_dict()["datasets"]["ogg"],
+            "dev": nn_cv_data_inputs["switchboard.cv"].get_data_dict()["datasets"]["ogg"],
+            "eval_datasets": {
+                "devtrain": nn_devtrain_data_inputs["switchboard.devtrain"].get_data_dict()["datasets"]["ogg"],
+            },
+        }
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:70]: You can always use the default (`else` branch) to create `returnn_datasets` and only `if use_multi_proc_dataset` then `returnn_datasets["train"] = {"class": ..., "dataset": returnn_datasets.pop("train")` or similar. That would be easier to read.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 20]
```diff
@@ -719,41 +724,354 @@ def run_scf_baseline():
...
     returnn_args = {
-        "batch_size": 5000,
+        "batch_size": 7000,
         "rasr_binary_path": RASR_BINARY_PATH,
         "rasr_loss_corpus_path": rasr_loss_corpus_path,
         "rasr_loss_corpus_segments": rasr_loss_corpus_segments,
         "rasr_loss_lexicon_path": rasr_loss_lexicon_path,
         "datasets": returnn_datasets,
         "extra_args": {
-            "accum_grad_multiple_step": 2,
```
[the remark this answers is not in the document]
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:729]: It seems unnecessary to change the position of this argument.

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 21]
```diff
@@ -918,13 +1136,11 @@ def py():
...
-            report_base,
-            report_mel,
-            report_scf,
-        ]
-    )
+    report = Report.merge_reports([
+        report_base,
+        report_mel,
+        report_scf,
+    ])
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:1143]: This looks like reverting the black changes. Please apply black to the file and check for these unnecessary changes.

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:232]: @Max-Ryujin this is still wrong, isn't it?

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 22]
```diff
@@ -47,15 +47,20 @@ def get_datasets(**kwargs):
         returnn_python_exe=RETURNN_EXE,
         **kwargs,
     )
-
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:50]: I think the empty lines give a bit more structure to the function's content, so I'd keep them.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 23]
```diff
@@ -47,15 +47,20 @@ def get_datasets(**kwargs):
...
         },
     }
-
+    if use_multi_proc_dataset:
+        returnn_datasets["train"] = {
+            "class": "MultiProcDataset",
+            "dataset": nn_train_data_inputs["switchboard.train"].get_data_dict()["datasets"]["ogg"],
+            "num_workers": 2,
+            "buffer_size": 5,
+        }
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:63]:
```suggestion
    
    if use_multi_proc_dataset:
        returnn_datasets["train"] = {
            "class": "MultiProcDataset",
            "dataset": nn_train_data_inputs["switchboard.train"].get_data_dict()["datasets"]["ogg"],
            "num_workers": 2,
            "buffer_size": 5,
        }
        
```

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 24]
```diff
@@ -734,6 +739,14 @@ def run_scf_baseline():
...
     }
     feature_args = {"class": "ScfNetwork", "size_tf": 256 // 2, "stride_tf": 10 // 2}
+    lr_args = {
+        "peak_lr": 4e-4,
+        "start_lr": 1.325e-05,
+        "end_lr": 1e-5,
+        "increase_epochs": 180,
+        "decrease_epochs": 180,
+        "final_epochs": 0,
+    }
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:749]: Is this used anywhere?

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py (hunk 14 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:887]: @Max-Ryujin This is still valid. Can't you e.g. create a list with all perturbations and then have just one loop through it? Something like
```
perturbations = [{"speed": [<list that you now have for speed>], "tempo": ...]
for perturbation in perturbations:
    ...
```
Then you can use `perturbation` instead of `audio_perturb_args` and just need a slight generalization of how you create the name.

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:232]: this is now dependent on `use_multi_proc_dataset`

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py (hunk 22 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:50]: even if black disagrees?

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py (hunk 24 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:749]: This was indeed a dublicate

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py (hunk 14 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:887]: That would make the code shorter, but maybe less understandable. There are also differences in the amount and names of parameters each perturbation needs, which would make the `WaveformPerturbation` call more complex

Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:883]: The code for each type of perturbation is exactly the same except for alias, report args and creation of `audio_perturb_args`. If you create a list of all perturbations where each entry is exactly what can be passed to `audio_perturb_args`, I do not think this is overly complicated or difficult to read. You just need a way to parse name and report args from that entry.

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 25]
```diff
@@ -776,6 +784,123 @@ def run_scf_baseline():
     return report
 
 
+def run_scf_audio_perturbation_gridsearch():
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:787]:
```suggestion
def run_scf_audio_perturbation():
```

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 26]
```diff
@@ -776,6 +784,123 @@ def run_scf_baseline():
...
+        {'preemphasis': {'prob': 0.9, 'minimum': 0.9, 'maximum': 1.0}},
+        {'preemphasis': {'prob': 0.9, 'minimum': 0.8, 'maximum': 1.0}},
+        {'preemphasis': {'prob': 0.8, 'minimum': 0.9, 'maximum': 1.0}},
+        {'preemphasis': {'prob': 0.8, 'minimum': 0.8, 'maximum': 1.0}},
+        {'codecs': [{'encoding': 'ULAW', 'prob': 0.4}]},
+        {'codecs': [{"encoding": 'ULAW', 'prob': 0.6}]},
+        {'non_linearity': {'prob': 0.4,  'minimum': 0.1, 'maximum': 0.3}},
+        {'non_linearity': {'prob': 0.4,  'minimum': 0.2, 'maximum': 0.4}},
+        {'non_linearity': {'prob': 0.6,  'minimum': 0.1, 'maximum': 0.3}},
+        {'non_linearity': {'prob': 0.6,  'minimum': 0.2, 'maximum': 0.4}},
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:846]: We usually used `"` in this file and not `'`. Please stay consistent in style.

More generally, let's discuss the exact perturbation parameters offline next time before merging and running all of this.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 27]
```diff
@@ -776,6 +784,123 @@ def run_scf_baseline():
...
+        arg_key = list(args.keys())[0]
+        arg_values = list(args.values())[0]
+
+        # Check if arg_values is a dictionary (has 'minimum' and 'maximum') or a list (like 'codecs')
+        if isinstance(arg_values, dict):
+            key_suffix = f"{arg_key}_{arg_values['prob']}_{arg_values['minimum']}_{arg_values['maximum']}_"
+            report_values = f"{arg_key}: '{arg_values['prob']}_{arg_values['minimum']}_{arg_values['maximum']}'"
+        elif isinstance(arg_values, list):
+            key_suffix = f"{arg_key}_{arg_values[0]['encoding']}_{arg_values[0]['prob']}_"
+            report_values = f"{arg_key} (encoding: {arg_values[0]['encoding']}): '{arg_values[0]['prob']}'"
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:861]: This does not work when combining perturbations. You could e.g. have `args =  {'speed': {'prob': 0.6, 'minimum': 0.8, 'maximum': 1.2},  'tempo': {'prob': 0.6, 'minimum': 0.8, 'maximum': 1.2}}`.  You could iterate through the `args` dict and create a name based on specific rules for the keys. E.g.
```
if key in ["speed", "tempo", ...]:
    name += f"{key}_{value["prob"]}_..."
elif key in ["codecs"]:
    name += ...
else:
    raise NotImplementedError
```
Maybe it also helps to create a small function to parse the args into a name. Similarly for the report args (could be done in the same helper function).

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 28]
```diff
@@ -776,6 +784,123 @@ def run_scf_baseline():
...
+
+        # Construct the key and report_args
+        key = f"scf_bs2x5k_perturb_{key_suffix}"
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:864]: `key` is a very generic variable name here. Maybe `exp_name` or similar would be more expressive?

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py (hunk 27 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py]: I made a helper function that generates both

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 29]
```diff
@@ -776,6 +784,144 @@ def run_scf_baseline():
...
+        {"preemphasis": {"prob": 0.8, "minimum": 0.8, "maximum": 1.0}},
+        {"codecs": [{"encoding": "ULAW", "prob": 0.4}]},
+        {"codecs": [{"encoding": "ULAW", "prob": 0.6}]},
+        {"non_linearity": {"prob": 0.4, "minimum": 0.1, "maximum": 0.3}},
+        {"non_linearity": {"prob": 0.4, "minimum": 0.2, "maximum": 0.4}},
+        {"non_linearity": {"prob": 0.6, "minimum": 0.1, "maximum": 0.3}},
+        {"non_linearity": {"prob": 0.6, "minimum": 0.2, "maximum": 0.4}},
+    ]
+
+    def process_args(args):
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:849]:
```suggestion
    def process_args(args: Dict[str, Any]):
```
Plus `from typing import Any, Dict` at the top.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 30]
```diff
@@ -776,6 +784,144 @@ def run_scf_baseline():
...
+        """
+        Process the argument dictionary to generate a key string and a report string.
+
+        Args:
+            args (dict): The argument dictionary to process.
+
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:855]:
```suggestion
```
I know it's not perfectly consistent in this file but recently we mostly used the above typing hints.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 31]
```diff
@@ -776,6 +784,144 @@ def run_scf_baseline():
...
+        Returns:
+            tuple: A tuple containing the key string and the report string.
+        """
+        arg_exp_name = list(args.keys())
+        arg_values = list(args.values())
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:860]:
```suggestion
```

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 32]
```diff
@@ -776,6 +784,144 @@ def run_scf_baseline():
...
+        key_string = ""
+        report_values = ""
+
+        for i in range(len(arg_exp_name)):
+            key = arg_exp_name[i]
+            value = arg_values[i]
+
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:867]:
```suggestion
        for key, value in args.items():
```

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 33]
```diff
@@ -776,6 +784,144 @@ def run_scf_baseline():
...
+            if key in ["speed", "tempo", "preemphasis", "non_linearity"]:
+                key_string += f"{key}_{value['prob']}_{value['minimum']}_{value['maximum']}_"
+                report_values += f"{key}: '{value['prob']}_{value['minimum']}_{value['maximum']}' "
+            elif key == "codecs":
+                key_string += f"{key}_{value[0]['encoding']}_{value[0]['prob']}_"
+                report_values += f"{key} (encoding: {value[0]['encoding']}): '{value[0]['prob']}' "
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:873]: Why only index 0? I'd argue if you have multiple codecs, you'd also want the alias to represent this.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 34]
```diff
@@ -776,6 +784,144 @@ def run_scf_baseline():
...
+            else:
+                raise ValueError(f"Unknown argument name: {key}")
+
+        return key_string, report_values
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:877]: Is it intended that `key_string` always has a trailing `_`?

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 35]
```diff
@@ -776,6 +784,144 @@ def run_scf_baseline():
...
+
+    nn_base_args = {}
+
+    for args in perturbation_args:
+        exp_name_suffix, report_values = process_args(args)
+
+        # Construct the exp_name and report_args
+        exp_name = f"scf_bs2x5k_perturb_{exp_name_suffix}"
+        report_args = {exp_name: report_values}
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:886]: Why is `exp_name` needed here? I think in the other experiments, it is not like this.

Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/experiments.py:886]: You could just take the function's output as `report_args` (and name it accordingly).

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 36]
```diff
@@ -776,6 +785,140 @@ def run_scf_baseline():
...
+        {"non_linearity": {"prob": 0.6, "minimum": 0.1, "maximum": 0.3}},
+        {"non_linearity": {"prob": 0.6, "minimum": 0.2, "maximum": 0.4}},
+    ]
+
+    def process_args(args: Dict[str, Any]):
+        """
+        Process the argument dictionary to generate a key string and a report string.
+
+        Returns:
+            tuple: A tuple containing the key string and the report string.
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:855]:
```suggestion
            tuple: A tuple containing the key string and the report dict.
```

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 37]
```diff
@@ -776,6 +785,140 @@ def run_scf_baseline():
...
+                key_string += key_component
+                report_values[key] = f"{value['prob']}_{value['minimum']}_{value['maximum']}"
+            elif key == "codecs":
+                codecs_str = "_".join([f"{codec['encoding']}_{codec['prob']}" for codec in value])
+                key_string += f"{key}_{codecs_str}_"
+                report_values[key] = codecs_str
+            else:
+                raise ValueError(f"Unknown argument name: {key}")
+
+        return key_string, report_dict
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:874]: Did you test this? You return `report_dict` here which does not seem to exist before.

**Review by vieting — CHANGES_REQUESTED** [review]

**Review by vieting — APPROVED** [review]
